32/berubejd/selenium_testing.py

```diff
@@ -0,0 +1,234 @@
+#!/usr/bin/env python3.8
+
+from pathlib import Path
+from selenium import webdriver
+from selenium.webdriver.common.keys import Keys
+from time import sleep
+
+
+def setup(url=None):
+    if not url:
+        url = "http://pyplanet.herokuapp.com/"
+
+    print("\tStarting browser setup...")
+
+    chrome_driver = Path.cwd() / Path("chromedriver.exe")
+    driver = webdriver.Chrome(executable_path=chrome_driver)
+    driver.get(url)
+
+    return driver
+
+
+def test_homepage(driver):
+    """ Test homepage header and nav links """
+
+    print("\tPerforming anonymous user homepage testing...")
+
+    # Header
+    header = driver.find_element_by_xpath("/html/body/header/h1")
+    assert (
+        "PyBites 100 Days of Django" in header.text
+    ), "Incorrect header text or text missing"
+
+    # Login link
+    login_link = driver.find_element_by_xpath('//*[@id="login"]/a[1]')
+    assert "Login" in login_link.text, "Incorrect login link or link missing"
+
+    # Home link
+    home_link = driver.find_element_by_xpath('//*[@id="login"]/a[2]')
+    assert "Home" in home_link.text, "Incorrect home link or link missing"
+
+
+def test_sharer_app(driver):
+    """ Test sharer app link and generated table """
+
+    print("\tPerforming anonymous user App testing...")
+
+    # Navigate to the Sharer App
+    driver.find_element_by_xpath("/html/body/main/ul/li/a").click()
+
+    # Collect table header and links
+    table_header = driver.find_element_by_xpath("/html/body/main/table/thead")
+    assert "Title" in table_header.text, "Incorrect table header"
+
+    table_links = driver.find_elements_by_xpath("/html/body/main/table/tbody/tr/td/a")
+    assert len(table_links) == 100, "Incorrect number of table rows"
+
+
+def test_article_link(driver):
+    """ Test table link links to proper article and the "Go back" button is the only one available 
+        and works as expected (in next step) """
+
+    print("\tPerforming anonymous user article testing...")
+
+    # Find and save the first article in the table
+    table_link = driver.find_element_by_xpath("/html/body/main/table/tbody/tr[1]/td/a")
+    link_title = table_link.text
+
+    # Navigate to that article
+    sleep(2)
+
+    table_link.click()
+
+    # Check that this is the correct article by comparing titles
```

Comment:
Might this be unnecessary if you used one of the Selenium Python driver's Wait functions?

Docs at:
https://selenium-python.readthedocs.io/waits.html

Reply:
I tried both implicit and explicit waits but explicit waits probably deserve some additional research into the best way to make this work rather than hardcoded sleep statements.

Reply:
I retested a vast number of wait related changes.  Strangely, in every case, the page was entirely rendered and all elements were being properly extracted from the DOM.  Input continued to fail without the sleep statements, however.

I implemented javascript input calls that worked without issue.  This may or may not be related to a bug somewhere in Selenium or the Chrome webdriver but, although there are a number of complaints, neither project is looking to address is.  The change I implemented is the only one recommended by the community that didn't require while loops and sleep statements.

I learned a lot more about Selenium waits, however. :)